src/Aspire.Dashboard/Components/Pages/TelemetryEnabledComponentBase.cs

```diff
@@ -0,0 +1,95 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Aspire.Dashboard.Telemetry;
+using Microsoft.AspNetCore.Components;
+using Microsoft.VisualStudio.Telemetry;
+
+namespace Aspire.Dashboard.Components.Pages;
+
+/// <summary>
+/// A base class for components that would like to opt-in to automatic telemetry events, capturing
+/// 1) initial render time
+/// 2) component properties on each render
+///
+/// <remarks>If overriding <see cref="OnInitializedAsync"/>, <see cref="OnParametersSetAsync"/>/<see cref="OnParametersSet"/>, or <see cref="OnAfterRenderAsync"/>/<see cref="OnAfterRender"/>,
+/// inheritors <b>must</b> the accompanying base methods for telemetry to work properly.</remarks>
+/// </summary>
+public abstract class TelemetryEnabledComponentBase : ComponentBase
+{
+    private ITelemetryResponse<StartOperationResponse>? _loadOperation;
+
+    [Inject]
+    public required NavigationManager NavigationManager { get; init; }
+
+    [Inject]
+    public required IAspireTelemetryService TelemetryService { get; init; }
+
+    protected abstract string ComponentId { get; }
+
+    protected virtual Dictionary<string, AspireTelemetryProperty> GetTelemetryProperties() => [];
+
+    // Use OnInitializedAsync instead so we ensure telemetry is initialized.
+    protected sealed override void OnInitialized()
+    {
+    }
+
+    protected override async Task OnInitializedAsync()
+    {
+        await TelemetryService.InitializeAsync();
+
+        if (TelemetryService.IsTelemetryEnabled)
+        {
+            var request = new StartOperationRequest(
+                TelemetryEventKeys.InitializeComponent,
+                new AspireTelemetryScopeSettings(new Dictionary<string, AspireTelemetryProperty> {
+                    // Component properties
+                    { TelemetryPropertyKeys.DashboardComponentId, new AspireTelemetryProperty(ComponentId) }
+                }));
+            _loadOperation = await TelemetryService.StartUserTaskAsync(request);
```

Comment:
This is adding a HTTP request to every page load. What is the performance impact?

I think physically sending telemetry needs to be taken out of inline dashboard operations. UI and performance of the app shouldn't be impacted by making HTTP requests.

Instead, telemetry events should be written by the UI to a channel, and then there is a background thread reads telemetry events from the channel and sends it to the telemetry backend.

Reply:
A channel makes sense

Reply:
I see that some methods return a response, e.g. `StartOperationResponse`, which have an ID on them. To support this with a channel, we should create our own ID (can just be an incrementing `long`) and return it from the method. Then there is an internal map between our ID and the ID eventually returned by the response.

Then when we end the operation, we pass in our ID, it's mapped to the response ID, and the external endpoint is called with the response ID.